examples/virtual-dom-wip/platform/Html/Internal.roc

```diff
@@ -0,0 +1,571 @@
+Html state : [
+    None,
+    Text MaybeJsIndex Str,
+    Element Str MaybeJsIndex Nat (List (Attribute state)) (List (Html state)),
+]
+
+MaybeJsIndex : [
+    NotRendered, # There's no JavaScript index for virtual nodes not yet rendered to the DOM
+    Rendered Nat, # Index of the corresponding real DOM node in the JavaScript `nodes` array
+]
+
+Attribute state : [
+    EventListener Str (List CyclicStructureAccessor) (MaybeRenderedHandler state),
+    HtmlAttr Str Str,
+    DomProp Str (List U8),
+    Style Str Str,
+]
+
+MaybeRenderedHandler state : [
+    NotRendered (Handler state), # Virtual DOM node, not yet rendered. Contains the lambda from app code.
+    Rendered Nat, # Index into a Roc List of lamdas. JS knows this index, and the List is refreshed with new lambdas on every render.
+]
+
+CyclicStructureAccessor : [
+    ObjectField Str CyclicStructureAccessor,
+    ArrayIndex Nat CyclicStructureAccessor,
+    SerializableValue,
+]
+
+# If we are only exposing the functions then are we better off just turning everything into a Custom?
+# At some point we need a common format anyway. Wrapper lambda is irrelevant for perf in context of an event.
+Handler state := [
+    Normal (state, List (List U8) -> Action state),
+    Custom (state, List (List U8) -> { action : Action state, stopPropagation : Bool, preventDefault : Bool }),
+]
+
+# -------------------------------
+#   VIEW FUNCTIONS
+# -------------------------------
+## Define an HTML Element
+element : Str -> (List (Attribute state), List (Html state) -> Html state)
+element = \tagName ->
+    \attrs, children ->
+        # While building the node tree, calculate the size of Str it will render to
+        withTag = 2 * (3 + Str.countUtf8Bytes tagName)
+        withAttrs = List.walk attrs withTag \acc, attr -> acc + attrSize attr
+        totalSize = List.walk children withAttrs \acc, child -> acc + nodeSize child
+
+        Element tagName NotRendered totalSize attrs children
+
+text : Str -> Html state
+text = \content -> Text NotRendered content
+
+none : Html state
+none = None
+
+nodeSize : Html state -> Nat
+nodeSize = \node ->
+    when node is
+        Text _ content -> Str.countUtf8Bytes content
+        Element _ _ size _ _ -> size
+        None -> 0
+
+attrSize : Attribute state -> Nat
+attrSize = \attr ->
+    when attr is
+        EventListener _ _ _ -> 0
+        HtmlAttr key value -> 4 + Str.countUtf8Bytes key + Str.countUtf8Bytes value
+        DomProp _ _ -> 0
+        Style key value -> 4 + Str.countUtf8Bytes key + Str.countUtf8Bytes value
+
+# -------------------------------
+#   STATIC HTML
+# -------------------------------
+appendRenderedStatic : Str, Html [] -> Str
+appendRenderedStatic = \buffer, node ->
+    when node is
+        Text _ content ->
+            Str.concat buffer content
+
+        Element name _ _ attrs children ->
+            withTagName = "\(buffer)<\(name)"
```

Comment:
betting this takes advantage of in-place mutation, but could be interesting to run some experiments later to make sure it does

Reply:
Yeah it's specifically written with that in mind! 
The precursor to this was the static site generator, and at the time I checked the Zig builtins to make sure they do the right thing. I did [this PR](https://github.com/roc-lang/roc/pull/3962) to get rid of some unnecessary allocations that were happening by making sure we use `capacity` properly.
But yeah it would be good to confirm it somehow! Mayne using a custom `roc_alloc` that counts allocations and reports them in a test. I did something like that in `test_gen` but this could be platform-specific.
Good idea!